probe_src/probe_frontend/python/probe.py

```diff
@@ -0,0 +1,35 @@
+
+import typing
+import json
+import subprocess
+from . import ops
+
+OpTable = typing.Mapping[int, typing.Mapping[int, typing.Mapping[int, typing.List[ops.Op]]]]
+
+def load_log(path: str) -> OpTable:
+    ret: dict[int, dict[int, dict[int, list[ops.Op]]]] = {}
+
+
+    lines = subprocess.run(
+        ["probe", "dump", "--json", "--input", path], 
+        capture_output=True, 
+        encoding="utf-8"
+    )
+    jsonlines = [json.loads(x) for x in lines.stdout.strip().split('\n')]
+
+    for item in jsonlines:
+        pid: int = item['pid']
+        epoch: int = item['exec_epoch']
+        tid: int = item['tid']
+        op: ops.Op = ops.Op(**item['op'])
```

Comment:
I think the more elegant way to do this would be to have every JSON object have a field called `_type`.

Then we supply `object_hook` to `json.loads` [docs](https://docs.python.org/3/library/json.html#json.load) (for an example, search for `__complex__`). Then we use `_type` to dispatch the constructor.

Then, we can use the default constructor for the dataclasses (objects in the fields of a dataclass would already get constructed by the time it got there).

Also, we don't have to hack around unions so much. The object gets sent to the right dataclass, by looking at `_type`, and the struct containing the union can have: `value: Mode | Ownership | Times`.

Once we're there, we might as well also set `@dataclass(frozen=True)`, which makes the classes immutable.

Reply:
> Then we supply object_hook to json.loads [docs](https://docs.python.org/3/library/json.html#json.load). Then we use _type to dispatch the constructor. This is also slightly more robust to changes in the structure, since the data carries a type indicator.
>
> Then, we can use the default constructor for the dataclasses (objects in the fields of a dataclass would already get constructed by the time it got there). Hopefully that simplifies the Rust code (don't have to generate __init__).

That does seem better, but I'm gonna need to take some time to figure out how to implement it (I have very little experience with typed python or rust proc-macros). Would the addition of a `_type` field apply only to top-level json objects (like `OpenOp`) or also child objects (like `Path`)?

> Also, we don't have to hack around unions so much. The object gets sent to the right union-variant by looking at _type, and the struct containing the union UpdateMetadataOp can have: value: Mode | Ownership | Times. Hopefully that also simplifies the Rust code.

Currently, each dataclass is produced in a way that totally blind of any others (other than the names of dataclasses already created); due to how rust proc-macros work, there's no way for me to lookup the variants of an enum from just it's name, this complicates inserting the type union directly into the struct containing that enum. An approach to that issue would be to define each enum variant as a dataclass and then define the enum itself as a type alias for a union across those dataclasses

The ideal approach would be to change how the python codegen is done so that the properties of each struct/enum occurs in its derive macro, and then a separate function-like macro takes the accumulated data and generates the file, this would have several benefits:

- solves the above issue (maybe, it would make dependency resolution much harder so it might still make sense to define type aliases)
- proper dependency resolver: currently, each dataclass can only be generated if all its child dataclasses have already been generated, this isn't something we should assume in rust since rust structs have no such requirement.
- the addition of a third macro could allow for additional, manually inserted, python code without interfering with the rest of the automatic codegen, this could solve the issue I mentioned with adding a `dirfd` property to `Path`.

Reply:
> Would the addition of a `_type` field apply only to top-level json objects (like `OpenOp`) or also child objects (like `Path`)?
The latter, all objects that are serialized as maps in JSON.

> there's no way for me to lookup the variants of an enum from just it's name
Just to confirm, you're saying it is difficult to go from `MetadataValue` to `{Path, Ownership, Times}` in the context of the struct-definition of `MetadataOp`?

> enum itself as a type alias for a union across those dataclasses
Here's the Python for that:
```py
MetadataKind: typing.TypeAlias = Path | Ownership | Times
```

Reply:
> Just to confirm, you're saying it is difficult to go from MetadataValue to {Path, Ownership, Times} in the context of the struct-definition of MetadataOp?

yeah, rust proc macros just do AST manipulation, so all the context you get is the part of the AST passed into the macro. For a derive macro that's the struct definiton it's attached to; all i can glean from a field of type `MetadataValue` in the definition of `MetadataOp` is that _some_ type with that identifier exists in scope.

> Here's the Python for that: ...

Thanks!

Reply:
> Then, we can use the default constructor for the dataclasses (objects in the fields of a dataclass would already get constructed by the time it got there). Hopefully that simplifies the Rust code (don't have to generate __init__).

One other thought I just had, I'm not 100% sure this will solve the whole issue, as I previously mentioned to you on matrix, `CString`s get serialized to arrays of numbers, the `bytes` constructor does the conversion to a byte string in python, and I'm not sure how that'd be handled by the default dataclass constructor.

Reply:
solved in https://github.com/charmoniumQ/PROBE/pull/20/commits/0beca52947f8d3c46cf527b04411f50ad76e39fe